dali/kernels/tensor_view.h

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_TENSOR_VIEW_
+#define DALI_KERNELS_TENSOR_VIEW_
+
+
+#include "shape.h"
+#include "backend_tags.h"
+
+namespace dali {
+
+template <StorageBackend backend, typename DataType, size_t dim_>
+struct TensorView {
+  DataType *data = nullptr;
+  TensorShape<dim_> shape;
+
+  static const size_t static_dim = dim_;
+  size_t dim() const { return ShapeDim(shape); }
+
+  template <typename Offset>
+  DataType *at(const Offset &pos) const {
+    return data + CalcOfffset(shape, pos);
+  }
+
+  template <typename... Indices>
+  DataType *at(Index idx0, Indices&& ...idx) const {
```

Comment:
A bit out of scope, but do we want to constrain number of idx passed here to `dim_`? At least in some cases.

Reply:
If we constrain it, then make it `sizeof...(idx)+1 <= dim_`, so we can easily slice. The proposed CalcOffset function already handles that.